### DISCUSSION TITLE
Using Zodios in a Monorepo

### DISCUSSION THREAD

**Discussion by epotter2297** [ecyrbe/zodios#212] [General]
I'm having some difficulties getting zodios to work in a monorepo correctly, so I set up a simple [example](https://github.com/epotter24G/zodios-router-error) here based on the turborepo starter project to demonstrate the problem. When I look at the router, I see that the type is `_zodios_core_lib_utils_types.IfEquals<Api, any, Omit<Router, "use"> & ZodiosUse<Context> & ContextRequestHandler<Context>, Omit<Router, any> & ZodiosHandlers<...> & ContextRequestHandler<...>>`

I'm not 100% sure if this is a bug, or something configured incorrectly in my tsconfig, so I wanted to check here before making an issue for it.

**Comment by ecyrbe** [comment]
the setup repository seems wrong : 
```json
{
  "name": "expresspkg",
  "version": "0.0.0",
  "description": "",
  "main": "index.ts",
  "module": "dist/index.mjs",
  "scripts": {
    "dev": "tsm index.ts"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/express": "^4.17.14",
    "@types/node": "^18.11.7",
    "eslint-config-custom": "workspace:*",
    "tsconfig": "workspace:*",
    "tsm": "^2.2.2",
    "typescript": "^4.7.4"
  },
  "dependencies": {
    "@zodios/core": "^10.4.3",
    "@zodios/express": "^10.4.0",
    "@zodios/openapi": "^10.4.2",
    "axios": "^1.1.3",
    "express": "^4.18.2",
    "zod": "^3.19.1"
  }
}
```
tsconfig is not setup in packages

**Reply by epotter2297** [reply]
Oh whoops, I had that at one point but I removed it while simplifying things. Just pushed up a change removing the reference (shouldn't be needed anymore).

I'm pretty sure it has something to do with the tsconfig, because if I delete it, it starts working. I'll keep looking and see if I can't narrow down what the problem is in the meantime.

**Comment by epotter2297** [comment]
@ecyrbe I think my problem doesn't have anything to do with monorepos at all, so I removed it from the example.

In the updated example, you can see that with the empty tsconfig (which should just load defaults), `req` is of type `any`. When you delete the empty `tsconfig.json`, it starts to work as intended again.

**Reply by ecyrbe** [reply]
Having an empty tsconfig would surely not work, because zodios needs to have strict mode activated.
  
But i have setup a mono repo with turbo repo and the issue i see is 'TS4023' error because of unamed exports.
  
It has to do how types are exported and the tsconfig options on the turborepo `base.json` file.
you can disable `declaration: true` and `declarationMap: true` and set them to false.

i've packaged a fix for the exports (`@zodios/express` v10.4.1), because i see unamed exports for `@zodios/express`, but the issue might also be on other packages.
i'll check a full setup.

**Comment by ecyrbe** [comment]
@epotter2297 so with `@zodios/express` v10.4.2 and the default turborepo setup, everything works ok.

here the web project package dependendencies:
```json
{
  "name": "web",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@tanstack/react-query": "^4.13.0",
    "@zodios/core": "^10.4.3",
    "@zodios/express": "^10.4.2",
    "@zodios/react": "^10.3.3",
    "axios": "^1.1.3",
    "express": "^4.18.2",
    "next": "12.3.2-canary.40",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "ui": "workspace:0.0.0",
    "zod": "^3.19.1"
  },
  "devDependencies": {
    "@babel/core": "^7.19.6",
    "@types/express": "^4.17.14",
    "@types/node": "^17.0.45",
    "@types/react": "^18.0.24",
    "@types/react-dom": "^18.0.8",
    "eslint": "7.32.0",
    "eslint-config-custom": "workspace:0.0.0",
    "tsconfig": "workspace:0.0.0",
    "typescript": "^4.8.4"
  }
}
```

**Comment by ecyrbe** [comment]
@epotter2297 fully working example here: https://github.com/ecyrbe/zodios-express/tree/main/examples/next-turbo

**Reply by epotter2297** [reply]
Perfect! I think that's everything sorted out, and I also found out that using `"moduleResolution": "NodeNext"` instead of `"moduleResolution": "Node"` was breaking some things for me too.
